mapswipe_workers/mapswipe_workers/firebase_to_postgres/update_data.py

```diff
@@ -2,93 +2,128 @@
+    return user_attribute_dict
 
-def update_user_data(user_ids: list = []) -> None:
+
+def update_user_data(user_ids: Optional[List[str]] = None) -> None:
     """Copies new users from Firebase to Postgres."""
     # TODO: On Conflict
     fb_db = auth.firebaseDB()
     pg_db = auth.postgresDB()
 
-    ref = fb_db.reference("v2/users")
-    last_updated = get_last_updated_timestamp()
-
-    if user_ids:
-        logger.info("Add custom user ids to user list, which will be updated.")
-        users = {}
-        for user_id in user_ids:
-            users[user_id] = ref.child(user_id).get()
-    elif last_updated:
-        # Get only new users from Firebase.
-        query = ref.order_by_child("created").start_at(last_updated)
-        users = query.get()
-        if len(users) == 0:
-            logger.info("there are no new users in Firebase.")
-        else:
-            # Delete first user in ordered dict (FIFO).
-            # This user is already in the database (user.created = last_updated).
-            users.popitem(last=False)
+    # get all user_ids from postgres
+    query = """SELECT user_id FROM users"""
+    postgres_user_info = pg_db.retr_query(query)
+    postgres_user_ids = []
+    for user_id in postgres_user_info:
+        postgres_user_ids.append(user_id[0])
+    logger.info(f"There {len(postgres_user_ids)} users in Postgres.")
+
+    if not user_ids:
+        # get all user_ids from firebase
+        firebase_user_ids = list(fb_db.reference("v2/users").get(shallow=True).keys())
+        logger.info(f"There {len(firebase_user_ids)} users in Firebase.")
     else:
-        # Get all users from Firebase.
-        users = ref.get()
-
-    for user_id, user in users.items():
-        # Convert timestamp (ISO 8601) from string to a datetime object.
-        try:
-            created = dt.datetime.strptime(
-                user["created"].replace("Z", ""), "%Y-%m-%dT%H:%M:%S.%f"
-            )
-        except KeyError:
-            # If user has no "created" attribute set it to current time.
-            created = dt.datetime.utcnow().isoformat()[0:-3] + "Z"
-            logger.info(
-                "user {0} didn't have a 'created' attribute. ".format(user_id)
-                + "Set it to '{0}' now.".format(created)
-            )
-        username = user.get("username", None)
-        query_update_user = """
-            INSERT INTO users (user_id, username, created)
-            VALUES(%s, %s, %s)
-            ON CONFLICT (user_id) DO UPDATE
-            SET username=%s,
-            created=%s;
-        """
-        data_update_user = [
-            user_id,
-            username,
-            created,
-            username,
-            created,
+        firebase_user_ids = user_ids
+
+    # Get difference between firebase users_ids and postgres user_ids.
+    # These are new users for which data is only available in Firebase so far.
+    new_user_ids = list(set(firebase_user_ids) - set(postgres_user_ids))
```

Comment:
This is very likely to be the case in most of the queries we do. We run this part of the code during the transfer or results and pass only the user ids that submitted results. So most of the time the list  of postgres_user_ids will be much longer than the firebase_user_ids list.

When a user deletes the Firebase account, we will still keep the user_id in the postgres table. We need it, since we use the user_id as a key for the results as well. However, most important we do not store the email of a user in postgres. The email is only in Firebase and will be deleted when you delete your account,

The list we get above will only show users that are not yet in postgres. If we want to get the opposite (only in postgres and not in firebase anymore) we would need to switch the order of the difference.

Reply:
understood, thanks! I guess I need to go reread the docs about python sets :)